radixdlt-java/src/main/java/com/radixdlt/client/core/ledger/InMemoryAtomStore.java

```diff
@@ -1,39 +1,56 @@
 package com.radixdlt.client.core.ledger;
 
+import com.google.common.collect.ImmutableSet;
 import com.radixdlt.client.core.atoms.Atom;
 import com.radixdlt.client.core.atoms.particles.Particle;
 import com.radixdlt.client.core.atoms.particles.Spin;
+import com.radixdlt.client.core.atoms.particles.SpunParticle;
 import com.radixdlt.client.core.ledger.AtomObservation.Type;
 import com.radixdlt.client.core.ledger.AtomObservation.AtomObservationUpdateType;
-import io.reactivex.subjects.Subject;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Objects;
-import java.util.Optional;
+import com.radixdlt.client.core.spins.SpinStateMachine;
+import io.reactivex.Observable;
+import io.reactivex.ObservableEmitter;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.stream.Stream;
 
 import com.radixdlt.client.atommodel.accounts.RadixAddress;
 
-import io.reactivex.Observable;
-import io.reactivex.subjects.ReplaySubject;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
-import org.radix.common.ID.EUID;
-
 /**
- * Implementation of a data store for all atoms in a shard
+ * An in memory storage of atoms and particles
  */
 public class InMemoryAtomStore implements AtomStore {
+	private final Map<Atom, AtomObservation> atoms = new ConcurrentHashMap<>();
+	private final Map<Particle, Set<Atom>> particleIndex = new ConcurrentHashMap<>();
+
+	private final Map<RadixAddress, CopyOnWriteArrayList<ObservableEmitter<AtomObservation>>> allObservers = new ConcurrentHashMap<>();
+	private final Map<RadixAddress, CopyOnWriteArrayList<ObservableEmitter<Long>>> allSyncers = new ConcurrentHashMap<>();
+
+	private final Object lock = new Object();
+	private final Map<RadixAddress, Boolean> syncedMap = new HashMap<>();
+
+	private void softDeleteDependentsOf(Atom atom) {
+		atom.particles(Spin.UP)
+			.forEach(p ->
+				particleIndex.get(p).forEach(a -> {
+					AtomObservation observation = atoms.get(a);
+					if (observation.getAtom().equals(atom)) {
+						return;
+					}
 
-	/**
-	 * The In Memory Atom Data Store
-	 */
-	private final ConcurrentHashMap<RadixAddress, Subject<AtomObservation>> cache = new ConcurrentHashMap<>();
+					if (observation.getUpdateType().getType() == Type.STORE || !observation.getUpdateType().isSoft()) {
+						// This first so that leaves get deleted first
+						softDeleteDependentsOf(observation.getAtom());
```

Comment:
Is this threadsafe? I realise you're using concurrent maps here, but I'm wondering if there is a case where we could end up in an invalid state because gathering and deletion is done in the same step.

On another note, what is the point of using concurrent maps (with the overhead they introduce) when all methods accessing them are `synchronised` anyway?

Reply:
Yes, good point, no need for concurrent maps now. But we shouldn't be getting into invalid state because deletes are just flagged rather than actually being removed from the map.